File: rustboyadvance-core/src/keypad.rs
#[derive(Debug, Primitive, PartialEq)]
#[repr(u8)]
pub enum Keys {
    ButtonA = 0,
    ButtonB = 1,
    Select = 2,
    Start = 3,
    Right = 4,
    Left = 5,
    Up = 6,
    Down = 7,
    ButtonR = 8,
    ButtonL = 9,
}

pub const NUM_KEYS: usize = 10;
pub const KEYINPUT_ALL_RELEASED: u16 = 0b1111111111;

#[derive(Debug, Primitive, PartialEq)]
#[repr(u8)]
pub enum KeyState {
    Pressed = 0,
    Released = 1,
}

impl Into<bool> for KeyState {
    fn into(self) -> bool {
        match self {
            KeyState::Pressed => false,
            KeyState::Released => true,
        }
    }
}

File: platform/android/app/src/main/java/com/mrmichel/rustboyadvance/Keypad.java
package com.mrmichel.rustboyadvance;

public class Keypad {
    private int keyState;

    public Keypad() {
        reset();
    }

    public void reset() {
        this.keyState = 0xffff;
    }

    public void onKeyDown(Key key) {
        this.keyState = this.keyState & ~(1 << key.keyBit);
    }

    public void onKeyUp(Key key) {
        this.keyState = this.keyState | (1 << key.keyBit);
    }

    public int getKeyState() {
        return keyState;
    }

    public enum Key {
        ButtonA(0),
        ButtonB(1),
        Select(2),
        Start(3),
        Right(4),
        Left(5),
        Up(6),
        Down(7),
        ButtonR(8),
        ButtonL(9);

        private final int keyBit;

        Key(int keyBit) {
            this.keyBit = keyBit;
        }
    }
}
